### PR TITLE
Prevent a shape being stored against sample pos

### LINKED ISSUE #28732: Add check to Instrument::markAsSamplePos to prevent shape being stored against SamplePos
The shape of the sample is supposed to be stored in the ExperimentInfo::m_sample member and not in the instrument description. To avoid confusion (eg the shape being stored in the wrong place or worse two different sample shapes being stored) add an error to markAsSamplePos to prevent an instrument component being labelled as the SamplePos if it also has a shape

This will also require a code change in InstrumentCreationHelper and ComponentCreationHelper which both set up a sample shape on the SamplePos instrument component. It will also require a change to AddAbsorptionWeightedPathLengthsTest which currently tries to work round this problem by copying the shape from the instrument component list to ExperimentInfo::m_sample

### PR DESCRIPTION
**Pull request by DannyHindson**
**Description of work.**

The shape of the sample should be stored against the m_sample field in the ExperimentInfo class. This PR prevents a second (potentially different) sample shape being stored in the instrument definition on the "SamplePos" component which should only have a position. It also removes a lots of instances of the "wrong" place being used - mostly in tests but also in ~8 algorithms.

I've achieved this by requiring the "SamplePos" component to be of type Component instead of ObjComponent (see change in Instrument.cpp). There were various places where the sample was set up to be an ObjComponent without it needing to be so I've changed these to Component.

At the last count there were 48 files changed in this PR - however 32 of them are tests or test helpers. And beyond that a lot of the changes are just to update a use of the ObjComponent class to the Component class. So the PR isn't as large as it appears. The key code changes are:
- Instrument.cpp, the main change to prevent an instrument component being marked as the sample being of a type that supports a shape
- InstrumentDefinitionParser.cpp, this unit now creates samples to be of type Component rather than ObjComponent
- change to bounding box logic in ComponentInfo.cpp and BoundingBox.cpp. As a result of this PR the sample is now a point-like object as far as the instrument definition is concerned. But it still needs to contribute to the bounding box calculations and I found some limitations in these two units that didn't handle point-like objects properly and resulted in them being ignored. Maybe this is partly why the sample had a shape set up in the instrument definition in lots of places

**To test:**

To a large extent there's no functional impact of this PR because it's main aim is to stop developers creating a sample shape in the wrong place (which results in that shape being ignored). So it's mainly a case of checking nothing's been broken which will be largely achieved by the unit\system tests. However two suggestions:

1) A quick test that will run through the changes to Instrument.cpp\InstrumentDefinitionParser.cpp is to load MAR11060.raw and check the "Instrument Viewer", "Show Detectors" options still work on the right click menu.

2) There is one small functional change to the CreateSampleWorkspace algorithm which now sets up a sample shape in the correct place. This means that running an absorption correction calculation (like the following) which analyses the sample shape will now work. It previously failed with an error saying "no sample shape" set up:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

ws=CreateSampleWorkspace()

ws_wave = ConvertUnits(ws,"Wavelength")

ws_corr = MonteCarloAbsorption(ws_wave)
```

Fixes #28732 . 

*This does not require release notes* because no external functional impact on user. It just stops Mantid developers setting up sample shapes in the wrong place that never do anything

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/TestHelpers/src/InstrumentCreationHelper.cpp [hunk 1]
```diff
@@ -81,10 +81,7 @@ void addFullInstrumentToWorkspace(MatrixWorkspace &workspace,
   instrument->markAsSource(source);
 
   // Define a sample as a simple sphere
```
Comment (ConorMFinn) [inline Framework/TestHelpers/src/InstrumentCreationHelper.cpp:83]: This comment may not be accurate any more?

**Review by cailafinn — CHANGES_REQUESTED** [review]
Just found the one comment that might need changing. Otherwise code looks good and changes function well.

File: Framework/TestHelpers/src/InstrumentCreationHelper.cpp (hunk 1 above)
Comment (DannyHindson) [reply Framework/TestHelpers/src/InstrumentCreationHelper.cpp]: Yes you're right. I've just updated it. Thanks for ploughing through this one!

**Review by cailafinn — APPROVED** [review]
Adding this restriction makes it much clearer where the sample shape should go. Code looks good and functions well. Nice work 👍

**Pull request merged**